### PR TITLE
[torch] [distributed] no need for distributed contract to ensure same order, can rely on FQN set being identical

### PR DESCRIPTION
**Pull request by colin2328**
Summary: as title. the prior contract is unnecessarily strict for composability contract

Differential Revision: D42117499

### CODE REVIEW

**Comment by awgu** [comment]
Just curious: What is the advantage of being able to change the parameter registration order?

**Comment by colin2328** [comment]
@awgu -> there is no 'advantage' per say, but its necessary for the current fbgemm implementation. Fbgemm will group tables together by sharding type, rather than by table_name and the order they were listed on non sharded nn.EmbeddingBag creation. (so, right now this is throwing  an error as I'm trying to integrate trec_shard with contract)

I don't believe this is necessary for composability at all.

File: torch/distributed/_composable/contract.py [hunk 1]
```diff
@@ -122,34 +122,25 @@ def wrapper(module: nn.Module, *args, **kwargs) -> Optional[nn.Module]:
                 f"nn.Module, but got {type(updated)}"
             )
 
-            def check_fqn(orig_fqns: List[str], new_fqns: List[str]):
+            def check_fqn(orig_fqns: Set[str], new_fqns: Set[str]):
                 if orig_fqns == new_fqns:
                     return
 
-                orig_fqn_set, new_fqn_set = set(orig_fqns), set(new_fqns)
```
Comment (mrshenli) [inline torch/distributed/_composable/contract.py:129]: Hey @colin2328, I think it is OK to drop this requirement, but it's not because this is not necessary. I feel we should keep the module's as close to its original behavior as possible, only drop the requirement if we have to. 

> there is no 'advantage' per say, 

There is definitely advantage here to keep the model's behavior intact. It can help avoid surprising users and avoid forcing them to spend additional time to debugging things. If we change that, we should clearly mention it in API's docstring. 

> but its necessary for the current fbgemm implementation. Fbgemm will group tables together by sharding type, rather than by table_name and the order they were listed on non sharded nn.EmbeddingBag creation. (so, right now this is throwing an error as I'm trying to integrate trec_shard with contract)

It is OK to make this temporary change to unblock the integration because fbgemm's current behavior. But I am curious is it technically necessary? Why do we have to abuse `nn.Module`'s existing API with a different behavior instead of using a new API to return the params/submodules with the grouped order? The same API should do the same thing if possible. 

> I don't believe this is necessary for composability at all.

It's not necessary for making composability work, but it is necessary for doing composability right.

Comment (awgu) [reply torch/distributed/_composable/contract.py:129]: Just an idea: Another option is for us to have a stricter contract for the PT-D APIs that specializes the general `@contract()` and enforces ordering, and TorchRec can opt out of that stricter one. The downside is that this may create some additional confusion to users since the contract nowv logically has an `if` / `else` branch.

For `fully_shard`, the existing unit tests enforce that `.named_parameters()` for a module annotated with `fully_shard` returns the same values in the same order as a non-annotated local module. I plan to keep that invariant regardless of if `@contract()` enforces that since it makes reasoning about the system simpler.

Comment (colin2328) [reply torch/distributed/_composable/contract.py:129]: @mrshenli , @awgu  thanks for the comments.

> It can help avoid surprising users and avoid forcing them to spend additional time to debugging things.

In trec ,sharding does 'module swapping' . we will swap an EBC with a shardedEBC (new module). Why do you thin kpotential changing order of params is a surprise? 

> we should clearly mention it in API's docstring.

Agreed, this is a good idea. I will add to trec_shard API docstring

> Why do we have to abuse nn.Module's existing API with a different behavior instead of using a new API to return the params/submodules with the grouped order? 

What exactly are you proposing? I mean, I guess we could manually reorder params inside named_parameters() override, is that what you mean? what would the benefit be?

>  but it is necessary for doing composability right.

I dont' follow this, why is keeping the order from unsharded -> sharded be 'doing composability right?

> For fully_shard, the existing unit tests enforce that .named_parameters() for a module annotated with fully_shard returns the same values in the same order as a non-annotated local module. I plan to keep that invariant regardless of if @contract() enforces that since it makes reasoning about the system simpler.

@awgu this sounds reasonable to me. IF we wanted we could pass this as flag to contract (i.e, enforce_param_order=True)

Comment (YLGH) [reply torch/distributed/_composable/contract.py:129]: I think it's reasonable to keep param order, I can imagine something doing something like

```
ebc = EmbeddingBagCollection([table_0, table_1])
...
opt_1 = SGD(next(ebc.parameters()))
```

and we should respect that even after sharding

We can control the order based on how we register the parameters, and the parameters will be in the same order as the table_name in the configs

Comment (mrshenli) [reply torch/distributed/_composable/contract.py:129]: > In trec ,sharding does 'module swapping' . we will swap an EBC with a shardedEBC (new module). Why do you thin kpotential changing order of params is a surprise?

TBH, module swapping itself might be a surprise. Or in general, any change to the original module's behavior might be a surprise to users, as we cannot anticipate how users gonna write their programs. So we prefer to only change original module's behaviors if we have to. 

> What exactly are you proposing? I mean, I guess we could manually reorder params inside named_parameters() override, is that what you mean? what would the benefit be?

What I am proposing is:

* Keep original modules behavior if we can
* If we have to change it due to technical reasons, we should clearly explain why we made this change. 

> I dont' follow this, why is keeping the order from unsharded -> sharded be 'doing composability right?

I guess this is just a philosophy that we don't change original module's behaviors if we don't have to. If the change is necessary, let's explain the technical reason for it. 

> @awgu this sounds reasonable to me. IF we wanted we could pass this as flag to contract (i.e, enforce_param_order=True)

Yep, that sounds good me. Shall we generalize this to sth like `strict=True` (please see if we can have a better name). This is not just for param order, we can also put other checks (e.g., returned module type) under this flag.
